api/controllers/tasks.py

```diff
@@ -1091,3 +1096,119 @@ def construct_trends_response_json(query_result):
         )
     else:
         return util.json_encode([])
+
+
+def get_secret_for_task_id(tid):
+    tm = TaskModel()
+    tups = (
+        tm.dbs.query(User.api_token)
+        .join(TaskUserPermission, TaskUserPermission.uid == User.id)
+        .filter(
+            db.and_(TaskUserPermission.type == "owner", TaskUserPermission.tid == tid)
+        )
+        .all()
+    )
+
+    ret = [tup[0] for tup in tups]
+    return ret
+
+
+@bottle.get("/tasks/listdatasets")
+def decen_eaas_list_datasets():
+    dm = DatasetModel()
+    tm = TaskModel()
+    req_data = bottle.request.json
+    data = req_data["data"]
+    task_code = data["task_code"]
+
+    # Get the TID for the task code
+    task = tm.getByTaskCode(task_code)
+
+    if not task:
+        # This means the task code the decen eval server
+        # started is wrong => return a useful message
+        bottle.abort(
+            400,
+            "That task code does not exist! "
+            "Please check eval_config.py to ensure your"
+            " task_code is correct.",
+        )
+
+    task_secrets = get_secret_for_task_id(task.id)
+
+    verified_across_keys = False
+    for task_sec in task_secrets:
+        if util.verified_data(req_data, task_sec):
+            verified_across_keys = True
+            break
+
+    if not verified_across_keys:
+        bottle.abort(401, "Operation not authorized")
+
+    # The only thing so far we need to update is the deployment_status
+    if set(data.keys()) != {"task_code"}:
+        bottle.abort(401, "Operation not authorized")
+
+    try:
+        # Get all datasets related to the task code
+        datasets = dm.getByTid(task.id)
+        json_encoded_datasets = []
+        for dataset in datasets:
+            json_encoded_datasets.append(util.json_encode(dm.to_dict(dataset)))
+
+        resp = {
+            "datasets_metadata": json_encoded_datasets,
+            "task_metadata": util.json_encode(tm.to_dict(task)),
+        }
+
+        return util.json_encode(resp)
+
+    except Exception as e:
+        logger.exception("Could not retrieve datasets for task: %s" % (e))
+        bottle.abort(400, "Could not retrieve datasets for task: %s" % (e))
+
+
+@bottle.get("/tasks/listmodelsids")
+def decen_eaas_list_model_ids():
```

Comment:
The difference for `listdatasets` is:
(1) security (as you mentioned), because decen has a non-cookie way of passing credentials (if you have the secret for your task — which you can only get from going to your task settings —, then you sign every packet we sent to prod with the secret, and prod verifies by retrieving the secret from DB and making sure the signatures match; this way only the true task owner can setup a working decen eval/build server)
(2) the response in `listdatasets` also includes some other task metadata that is required by the decentralized eval server in order to run; I was going to use this method but I didn't want to break anything in normal prod 

For `listmodelids` I think the only difference is security. 

I think it's better to leave the decen security, so we know that there is at least some level of security to all the decen endpoints. However, I do think we can combine the logic into one endpoint (just passing a flag to the endpoint if the response needs to change slightly, and to change how to verify the request), I can make an issue for that

Reply:
Ok sounds good, thanks!